### ISSUE TITLE
Does sympy have a functor for category theory module?

### ISSUE THREAD

**Issue by enjoysmath** [sympy/sympy#17964]
Question in the title.

How would I got about implementing one if there is not one?

Thanks.

**Comment by asmeurer** [comment]
The only mention of functors in the code is here, which says they aren't implemented yet https://github.com/sympy/sympy/blob/2662c098cb19fb48c821ca890d6d664a2efda555/sympy/categories/__init__.py#L5-L6. 

I don't know how to implement it. Maybe @scolobb has thoughts.

**Comment by enjoysmath** [comment]
@scolobb well, it's sort of like a function, so you could maybe inherit from that, except it has category as domain, codomain, and the call function on it checks whether input is a morphism or an object.  Etc.  If you get me set up, I could actually implement it.

**Comment by asmeurer** [comment]
I haven't used the category theory module much, but am I correct that it only supports finite categories? If so, then I would suggest just having it be an explicit map, which keeps track of every object and morphism mapping.  Presumably it only needs to actually keep track of a subset of the full mapping because it can use the fact that it commutes (is it enough to just store the mapping on morphisms?). 

If the module can represent infinite categories, then for Functor I would start with just an abstract class that doesn't do anything. I don't see anything about infinite categories in the code, but I might be missing it.

**Comment by enjoysmath** [comment]
> I haven't used the category theory module much, but am I correct that it only supports finite categories? If so, then I would suggest just having it be an explicit map, which keeps track of every object and morphism mapping. Presumably it only needs to actually keep track of a subset of the full mapping because it can use the fact that it commutes (is it enough to just store the mapping on morphisms?).
> 
> If the module can represent infinite categories, then for Functor I would start with just an abstract class that doesn't do anything. I don't see anything about infinite categories in the code, but I might be missing it.

Yes, any implementation of category theory by definition has to support the infinite.  I think the finiteness in the code is only because it works with "known assertions / objects" that are given by the user, and those by definition are finite.  So for instance the system knows that the user asserts that arrows f,g,h are in some abstract category C.  So it keeps track of those, presumeably for checking if things commutue, etc.

**Comment by asmeurer** [comment]
So what should the Functor object actually store? Or should it just be an abstract symbolic object that doesn't actually compute anything?

**Comment by enjoysmath** [comment]
A functor could have a method that takes as input a diagram in its domain
category, and output a corresponding diagram in the codomain.

This question comes up.  Do you store F(f) as a "composed symbol" (some
type I'm not aware of within SymPy) or just as Ff = Object("F(f)")?

**Comment by asmeurer** [comment]
SymPy objects are expression trees. Take a look at https://docs.sympy.org/latest/tutorial/manipulation.html. You would just have `F = Functor(f)` and `F(f)` where `f` is the category (presumably functors should map categories, not diagrams). 

You can easily build just that with

```
class Functor(Basic):
    pass
```

The question is, what other features should it have? Or should it just be that?

**Comment by scolobb** [comment]
Hi @enjoysmath !  Thank you for your interest in this stuff!  I'm somewhat busy right now, but I'll try to get back over the next couple days to discuss the subject.  Feel free to ping me if I don't :slightly_smiling_face:

**Comment by enjoysmath** [comment]
I said in my previous post, that it should have a diagram image taker
method.

There are many other things it should do!  It is a functor!  They're
everywhere in math.

**Comment by enjoysmath** [comment]
Okay, will do.

I am probably not going to develop for SymPy, but I'd like to extend it
nonetheless in my own application.

I will definitely need a Functor class.   Eventually I'll be writing more
concrete examples of categories, for example R-mod where R is a ring.

For that I will use more and more of other parts of SymPy.

I will implement a functor class, then a subclass that communicates with my
graphics.

Thank you all for your development work on SymPy!

**Comment by enjoysmath** [comment]
> Hi @enjoysmath ! Thank you for your interest in this stuff! I'm somewhat busy right now, but I'll try to get back over the next couple days to discuss the subject. Feel free to ping me if I don't 🙂

How do I ping you here?  Like this?:

@scolobb

**Comment by scolobb** [comment]
@enjoysmath Yeah, that works just fine :smile: I'll try to come on my own within a few days though and meaningfully contribute to your work.  It's quite exciting :blush:

**Comment by enjoysmath** [comment]
That would be great, because I have no idea how to connect it with the rest
of SymPy.  So let's both come up with some codes, mon. :D

**Comment by enjoysmath** [comment]
> @enjoysmath Yeah, that works just fine 😄 I'll try to come on my own within a few days though and meaningfully contribute to your work. It's quite exciting 😊

It is.  We are on the forefront of a great technological revolution in the way we do diagram chases.  With SymPy, you can even perhaps one day compute homology etc. :D
